Pretty print AST with color information

aisipos opened the pull request:
Fixes #168 

I should say first, this is probably a larger PR than you were expecting. I'm totally fine if we abandon this approach and go in a different direction, but I wanted to see what directions I could come up with and if any are useful.

The tl;dr currently is, if you have `foo.spy`:
```
def main(x: i32) -> i32:
    return x + 2 * 3
```
Then run:
```
spy --redshift --colorize foo.spy
```
You will see:
<img width="532" height="524" alt="image" src="https://github.com/user-attachments/assets/11536e5f-0ac7-4e85-b79e-78aafc7e1836" />

This is probably not quite what we were looking for, for one I suspect we don't want the delemiters like () [] colored, or things like the top level Module colored, but the approach is close enough that I wanted to get some feedback first before proceeding in this direction.

I tried to make small atomic commits introducing single concepts at a time:
- In 8c5022b0929af6791bff67cdba394d0fda06bd11 I tried to follow your advice in saving the pre-redshift color in `DopplerFrame` instances. The problem I ran into here is that `DopplerFrame` instances are essentially ephemeral, they are not saved anywhere that is easily retrieved after redshifting. 
- So instead in cbd96b46b724d10a3359fb0eaedc9b2037cc4c51 I added a `color` attribute to `Expr` nodes, and save the original color during the redshifting process. These are the instances that are used in the AST itself that gets pretty printed, so it seemed like the natural place to add them for this use case. I realize this attribute is not used in the actual compilation pipeline, so if we feel this means the methodology here is no good I'm fine with figuring out an alternate method.
- The way that we want to do syntax coloring in this new "mode" is different than before, so to make this work I introduced a concept of `color_mode` in ddc30180766dad413f5c9d0630c1bd3314f7f4ae. This way we can use different coloring logic depending on the `color_mode`.
- To avoid conflation of names and concepts, I separated out the concept of `text_color` from `color` in 1151a7d9700e1f670bf2f430be620eea3cec8d19. When I think "color" I think "red or blue", but "text_color" can be anything, even including "green" or "turquoise".
- I have several commits that made several new ways of setting a new `text_color` that I ultimately all abandoned in favor of 0e9e4596acc6eb6d072a7eee24eee0cf5a9fc6b3, which introduces an `active_text_color` attribute to `TextBuilder`. This feature gets used by the logic now enabled in the new `redshift` `color_mode`, so that multiple nodes and children can automatically get the parent color, as implemented in 0e9e4596acc6eb6d072a7eee24eee0cf5a9fc6b3. However, as noted above, this method might be too broad, so we can work on this before we possibly merge.

This PR will need at the very least some cleanup, squashing, and tweaking before it is ready to merge, which I'm happy to do. Or we might decide the approach is not useful and do something else entirely. But since it does at least come close to the original intent of #168 I wanted to put it up to get some feedback first.

antocuni commented:
Thank you a lot for your work, I love it! I especially appreciate the thoughtfulness you put in explaining things commit-by-commit, it made it much easier to review the code.

Let's deal with things one by one.

---

### `ast.Expr.color`

I can see why it looks appealing to do that, but I'd highly prefer not to, for two reasons:
1. `--colorize` is something that people will do only rarely, and I don't want to pay the memory cost for an extra field which is completely unused 99% of the time

2. I'm trying to treat `ast.*` classes are immutable. This will make it easier to switch to a more compact representation later (where e.g. all the AST is binary-encoded in a cache-friendly contiguous buffer), which will be important for the performance of the interpreter

I see the problem with DopplerFrame being ephemeral. So, here is a new suggestion: we could introduce a new field `vm.expr_color_map`, which maps `ast.Expr` to their `Color`.
In normal mode, `vm.expr_color_map` is None, but when `--colorize` is passed, we set it to a dictionary, and then DopplerFrame knows when/where to store the info.

Finally, we will pass `expr_color_map` to `ast_dump.Dumper` so that it knows the color of the nodes.

---

### TextBuilder improvements

I think that your approach of using just one `active_text_color` doesn't work in case of nested `color_block()`. Moreover, I'm not sure I like the fact that just entering the color_block writes to the terminal, because e.g. it breaks if you then call `write(s, color='...')` (I think, I admit I didn't try).

I tried a slightly different approach in PR #200 which I think is more general:
  - you can have nested `color()` blocks
  - you now have support for bg colors
  - you can always temporarily override color and bg inside individual `write` calls, so that e.g. the following works:
https://github.com/spylang/spy/blob/9adbd94ca966320a1fa8dca54cc4138842903537/spy/textbuilder.py#L265-L275

---

### Appearance of the colored AST

When it comes to aesthetic things are very personal and opinionated of course, but I find the "all blue all red" screenshot a bit "too much" :).

Some random notes:
  - not all nodes are expressions, and thus not all should be colored
  - the current pretty-print logic prints node names in blue, but that's just an accident. It probably makes more sense to use a different color (such as turquoise, same as py_ast) just to avoid confusions

I randomly experimented with two "low-key" approaches. I pushed my code in the branch `tmp-anto-ast-color`, which is based on your commit 743a008d.

Consider this example:
```python
def fn(x: int, y: int) -> int:
    return x + y + 2*3
```

The first is to colorize *only the node name* in red/blue (and use turquoise for "default" color for non-expr nodes). You can find the code in commit c010af25):
<img width="773" height="1221" alt="image" src="https://github.com/user-attachments/assets/aad153b8-2a4c-4ea7-bf69-3580301f4904" />

The second is to keep all foreground colors normal, and use background color to distinguish red/blue. You can find the code in commit a07f5ad4:
<img width="788" height="1247" alt="image" src="https://github.com/user-attachments/assets/7d41a96f-2db3-4972-9ce7-5ebf2d6bf013" />

I think I have a slight preference for the first, but I'm happy to hear opinions.

aisipos commented:
No worries, thanks for all the feedback! I review many PRs myself as well, so I also appreciate the value of making readable and reviewable PRs. Gotta write code to be readable! 🙂

Regarding the points you raised:
- OK, I'll abandon `ast.Expr.color` method as you noted. I was suspicious it would introduce some overhead, but since it did seem like the most direct route I thought I'd try it first to see how well it worked.  I'll instead try the `vm.expr_color_map` you suggested. 
- I saw that you merged a separate `TextBuilder` color PR. I'll rebase this PR on top of main and make use of the new features, but I might still try to preserve the naming of `text_color`. That way, if we grep in the codebase for `text_color` we'll see all the parts of the code that are trying to produce colored output. If we just use `color`, it becomes conflated with the code that detect the different node types.
- Regarding the proposed two styles of colorization you proposed, my initial reactions are
  - In favor of (2), 2 does draw the eye more than 1, in the additional use of background color as well as foreground color.
  - In favor of (1), 2 feels a bit "louder" than 1, in that more of the text is colored than 1.
- With that in mind, I might propose that we use a hybrid of the two, in that like (1) we only color the node's name, but we use the background color to highlight it like we do in (2). The nice part of this approach is that in this case foreground text_color corresponds to a node's "type", and the background text_color refers to a node's "color", so we are using a more consistent conceptual presentation. 
- We might not need the concept of `color_mode` any more, since if we have a `vm.expr_color_map` instance we can use it in the coloring, or if it is None we do nothing. The only real difference then is that the cli `--colorize` argument tells us to print the pre-redshifted AST, not the post redshift AST, as it currently does in this PR as of this writing.

Let me know what you think, I'll give this a try but probably will have a look at #173 first.

antocuni commented:
> I saw that you merged a separate TextBuilder color PR. I'll rebase this PR on top of main and make use of the new features, but I might still try to preserve the naming of text_color. That way, if we grep in the codebase for text_color we'll see all the parts of the code that are trying to produce colored output. If we just use color, it becomes conflated with the code that detects the different node types.

yes, that's a good idea. Other possible choices are `fgcolor` or `fg` (since we already have a `bg` attribute). But `text_color` is also good.

> With that in mind, I might propose that we use a hybrid of the two, in that like (1) we only color the node's name, but we use the background color to highlight it like we do in (2). The nice part of this approach is that in this case foreground text_color corresponds to a node's "type", and the background text_color refers to a node's "color", so we are using a more consistent conceptual presentation.

sounds good!
Also, eventually I'd like to be able to colorize also the source code (not only the AST), so this style plays well with that as well (e.g., we will be able to pretty-print the source code with red/blue background, and it will be consistent with the AST view).

> We might not need the concept of color_mode any more, since if we have a vm.expr_color_map instance we can use it in the coloring, or if it is None we do nothing. The only real difference then is that the cli --colorize argument tells us to print the pre-redshifted AST, not the post redshift AST, as it currently does in this PR as of this writing.

+1

aisipos commented:
OK, I essentially redid the PR with the new approach you suggested. I also made the cli work with just the `--colorize` option, there's no need now to run `--redshift --colorize`. I reverted all the earlier work in commit 879eb0926d9fd7d401b7a47678596adc429ddc7a, the commits after that add the new approach step by step. If we decide we like this approach, I'll delete all the old commits before we merge.

I added a new variable `expr_color_map` to `SPyVM`.  The main refactor was making this available to all the pretty print functions in `ast_dump.py`. I did this by adding an optional `vm` argument to `Dumper` and making it an instance attribute. I had to also add an optional `vm` argument to several functions that end up creating a Dumper instance, in particular `Node.pp`.

I tried my idea of using Blue / Red as the background color, and the current `text_color` concept as the foreground color, which partly looks like this:
<img width="297" height="188" alt="image" src="https://github.com/user-attachments/assets/92a1fd19-c33f-4759-809e-592f321c6c85" />

I have slightly mixed feelings about it now that I can see it working in a real output, the contrast between the foreground color and background color clashes a bit and lowers the legibility somewhat. However, the red and blue colors do stand out more than if they were just foreground colors. Perhaps if we just ignored` text_color` when `bg_color` is specified it would look better, but then we do lose some information. 

Let me know what you think about the current implementation. Happy to make more changes if it's not a good fit.

antocuni commented on spy/ast_dump.py:
```diff
@@ -1,23 +1,29 @@
          use_colors: bool = True,
          fields_to_ignore: Any = (),
          hl: Any = None,
+         vm: 'SPyVM | None' = None,
```
nitpick: so far I always used `Optional[T]` instead of `T | None`.
It's a very minor thing but I think there is value in being consistent.
(I _might_ be convinced that it's better to use `| None` instead of `Optional`, but then it's a topic for a different PR).

antocuni commented on spy/ast_dump.py:
```diff
@@ -76,8 +87,15 @@ def is_complex(obj: Any) -> bool:
+        # as the text background color
+        bg_color = None
+        if self.vm and self.vm.expr_color_map:
+            color: Color | None = self.vm.expr_color_map.get(node, None)
```
+0.5 to pass around `vm` if it helps, but note that we could also pass just `expr_color_map`, and cut the circular dependency between `ast.py` and `vm.py.`
Anyway, not a big deal.

aisipos replied:
Yeah, I decided to pass in `vm` since as more features get added, I didn't want to pass more and more arguments to the various pretty print functions. Then again, we probably don't want `vm` to become a "dumping ground" of global variables. But at least for now, passing in `vm` seemed like a good compromise.

antocuni commented on spy/cli.py:
```diff
@@ -55,6 +55,14 @@ class Arguments:
+            "-C", "--colorize",
+            help="Output the pre-redshifted AST with blue / red text colors."
+        )
+    ] = False
```
`-C` is already used as an alias for `--cwrite`.
I think that `--colorize` is more important, so +1 to use `-C` here, but we should remove `-C` from the `cwrite` option.

antocuni commented on spy/tests/test_doppler.py:
```diff
@@ -294,3 +294,41 @@ def foo() -> None:
+
+    def test_expr_color_map_populated(self):
+        # Verify that when vm.expr_color_map is not None, redshift populates it
+        from unittest.mock import patch
```
I'd use pytest's `monkeypatch` fixture if possible.

aisipos replied:
Done.

antocuni reviewed (approved):
I like it, thank you!
I have commented a couple of nitpicks, but nothing fundamental.
For the color yes, I agree that turquoise-on-blue and turquoise-on-red is not very visible, +1 to just use `default-on-bg` is bg is set.

aisipos commented:
@antocuni OK, I addressed all of your feedback, deleted all the old commits, and rebased on top of current `main`. The output now only sets the background color for red / blue:
<img width="300" height="186" alt="image" src="https://github.com/user-attachments/assets/74de2d08-7ce1-41c6-85b4-03463a5bc66d" />
Hopefully this fixes all our remaining issues.

antocuni commented:
@aisipos merged, thank you :pray:
